### PR TITLE
Improvements for iwyu_tool.py on Windows

### LINKED ISSUE #582: iwyu_tool.py source file filtering is not case-insensitive
On Windows (and other case-insensitive platforms), provided source paths must match the compilation database exactly, or they won't be processed and a warning will be printed.

Found by @jru here: https://github.com/include-what-you-use/include-what-you-use/pull/580#issuecomment-414697422

### LINKED ISSUE #583: iwyu_tool.py does not work on Windows
After PR #580, iwyu_tool.py no longer works on Windows.

This is because `shlex.split` does not handle Windows path delimiters and quoting correctly, so the command-line becomes jumbled.

### PR DESCRIPTION
**Pull request by kimgr**
* Attempt to fix #583 with custom command-string splitting
* Fix #582 by improving selection filtering
* Improve MSVC CL-style driver detection to be case-insensitive and also recognizing `clang-cl`

I've had no opportunity to test this :-/ @jru, could you take it for a spin?

### CODE REVIEW

File: iwyu_tool.py [hunk 1]
```diff
@@ -91,6 +91,73 @@ def clang_formatter(output):
...
+    # shlex.split with posix=False is known to not work perfectly for Windows
+    # command-lines, but it seems good enough in the general case for
+    # compilation databases.
+    windows = sys.platform.startswith('win')
+    cmd = shlex.split(cmdstr, posix=not windows)
+
+    # It does, however, retain quotes around arguments, which causes
+    # double-quoting and escaping when passed on to subprocess. Strip them.
+    if windows:
+        cmd = [strip_quotes(a) for a in cmd]
```
[the remark this answers is not in the document]
Comment (kimgr) [reply iwyu_tool.py]: I don't think it does break. I looked through subprocess.py and saw that on Windows, the first thing that happens is that the `args` list is passed to `list2cmdline`, which joins the arguments back to a string and quotes them if necessary. As part of that quoting, though, it escapes any existing quotes.

So an argument like `"C:\my file\with spaces.txt"` would turn into `"\"C:\my file\with spaces.txt\""`, which shouldn't be found.

I think it might make sense to add some of this rationale to the comment, though, I see now that I didn't mention subprocess there.

File: iwyu_tool.py [hunk 2]
```diff
@@ -91,6 +91,73 @@ def clang_formatter(output):
...
+
+    if not path.startswith(parent):
+        return False
+
+    # Now we know parent is a prefix of path, but they only share lineage if the
+    # difference between them starts with a path separator, e.g. /a/b/c/file
+    # is not a parent of /a/b/c/file.cpp, but /a/b/c and /a/b/c/ are.
+    parent = parent.rstrip(os.path.sep)
+    suffix = path[len(parent):]
+    return suffix.startswith(os.path.sep)
```
[the remark this answers is not in the document]
Comment (kimgr) [reply iwyu_tool.py]: I didn't want to touch the filesystem, only string manipulation.

File: iwyu_tool.py [hunk 3]
```diff
@@ -91,6 +91,73 @@ def clang_formatter(output):
...
+
+def is_msvc_driver(compile_command):
+    """ Return True if compile_command matches an MSVC CL-style driver. """
+    compile_command = normcase(compile_command)
+
+    if compile_command.endswith('cl.exe'):
+        # Native MSVC compiler or clang-cl.exe
+        return True
+
+    if compile_command.endswith('clang-cl'):
```
[the remark this answers is not in the document]
Comment (jru) [reply iwyu_tool.py]: True! :D

File: iwyu_tool.py [hunk 4]
```diff
@@ -91,6 +91,73 @@ def clang_formatter(output):
...
+else:
+    def normcase(s):
+        return s
+
+
+def is_msvc_driver(compile_command):
+    """ Return True if compile_command matches an MSVC CL-style driver. """
+    compile_command = normcase(compile_command)
+
+    if compile_command.endswith('cl.exe'):
```
[the remark this answers is not in the document]
Comment (jru) [reply iwyu_tool.py]: Yes, I also think is not likely to happen. Fine with leaving it for now.

File: iwyu_tool.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jru) [reply iwyu_tool.py]: I think I know what's going on now. The problem is not in the strip_quotes() function, but in shlex.split(.., posix=False). It splits incorrectly when a quoted argument with a space is appended to an unquoted one (|**something"with spaces"**| becomes |**something"with**|**spaces**|):

To verify this, I added an include directory with a space to the IWYU's root path and and to CMakeLists.txt:
```cmake
target_include_directories(include-what-you-use PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/inc\ ludes")
```

After this, the generation step will add this include path for each command in the compile_commands.json file:
```
-I"..\\tools\\clang\\tools\\iwyu\\inc ludes"  
```

When I run it under iwyu_tool.py I get this as the first line of output:
```
error: no such file or directory: 'ludes"'
```

So, in short, directories with spaces will not work on Windows. :(

**Comment by kimgr** [comment]
Thanks, looks like we're on the same page. Can you unblock the review state?

**Comment by jru** [comment]
Hi, I think this is still broken due to what I mentioned above: https://github.com/include-what-you-use/include-what-you-use/pull/584#discussion_r212985102

**Comment by kimgr** [comment]
Aargh, Github's notifications don't come through for code review comments, so I had missed that. Thanks, I'll look into it again.

**Comment by kimgr** [comment]
Closing this; I'm about to post fixes for all (:crossed_fingers:) Windows issues as separate pull requests.
